processing/src/main/java/io/druid/segment/IndexMerger.java

```diff
@@ -271,19 +274,52 @@ public File merge(
     );
   }
 
+  private List<String> getLongestSharedDimOrder(List<IndexableAdapter> indexes)
+  {
+    int maxSize = 0;
+    Iterable<String> orderingCandidate = null;
+    for (IndexableAdapter index : indexes) {
+      int iterSize = Iterators.size(index.getDimensionNames().iterator());
+      if (iterSize > maxSize) {
+        maxSize = iterSize;
+        orderingCandidate = index.getDimensionNames();
+      }
+    }
+
+    if (orderingCandidate == null) {
+      return null;
+    }
+
+    for (IndexableAdapter index : indexes) {
+      Iterator<String> candidateIter = orderingCandidate.iterator();
+      for (String matchDim : index.getDimensionNames()) {
+        boolean matched = false;
+        while (candidateIter.hasNext()) {
+          String nextDim = candidateIter.next();
+          if (matchDim.equals(nextDim)) {
+            matched = true;
+            break;
+          }
+        }
+        if (!matched) {
+          return null;
+        }
+      }
+    }
+    return ImmutableList.copyOf(orderingCandidate);
+  }
+
   private List<String> getMergedDimensions(List<IndexableAdapter> indexes)
   {
     if (indexes.size() == 0) {
       return ImmutableList.of();
     }
-    Indexed<String> dimOrder = indexes.get(0).getDimensionNames();
-    for (IndexableAdapter index : indexes) {
-      Indexed<String> dimOrder2 = index.getDimensionNames();
-      if(!Iterators.elementsEqual(dimOrder.iterator(), dimOrder2.iterator())) {
-        return getLexicographicMergedDimensions(indexes);
-      }
+    List<String> commonDimOrder = getLongestSharedDimOrder(indexes);
+    if (commonDimOrder == null) {
+      return getLexicographicMergedDimensions(indexes);
```

Comment:
Perhaps do a `log.warn` here, as this case is not expected during normal indexing.